pkg/pillar/agentbase/agent.go

```diff
@@ -15,54 +15,114 @@ import (
 // this interface.
 type AgentBase interface {
```

Comment:
This solution requires every microservice to implement these 3 methods and for the most part they are only copy-paste.
I would suggest instead to make a default implementation of AgentBase that each microservice would embed and override as needed. Better then to explain is to give an example (in the form of UT that can be run):

```
package agentbase_test

import (
	"flag"
	"fmt"
	"os"
	"testing"
)

// agentbase.go

type Agent interface {
	GetAgentBase() *AgentBase
	AddAgentSpecificCLIFlags(*flag.FlagSet)
	ProcessAgentSpecificCLIFlags(*flag.FlagSet)
}

type AgentOpt func(*AgentBase)

type AgentBase struct {
	agentName string
	arguments []string
        // etc.
}

func WithArguments(arguments []string) AgentOpt {
	return func(ab *AgentBase) {
		ab.arguments = arguments
	}
}

func Init(a Agent, agentName string, opts... AgentOpt) error {
	// this would do what currently is done by `InitAgentBaseContext` + `agentbase.Run()` functions
	ab := a.GetAgentBase()
	ab.agentName = agentName
	for _, opt := range opts {
		opt(ab)
	}
	flagSet := flag.NewFlagSet(ab.agentName, flag.ExitOnError)
	a.AddAgentSpecificCLIFlags(flagSet)
	if err := flagSet.Parse(ab.arguments); err != nil {
		return err
	}
	a.ProcessAgentSpecificCLIFlags(flagSet)
	return nil
}

func (a AgentBase) GetAgentBase() *AgentBase {
	return &a
}

func (a AgentBase) AddAgentSpecificCLIFlags(*flag.FlagSet) {
	fmt.Println("Default (NOOP): AddAgentSpecificCLIFlags")
}

func (a AgentBase) ProcessAgentSpecificCLIFlags(*flag.FlagSet) {
	fmt.Println("Default (NOOP): ProcessAgentSpecificCLIFlags")
}


// then inside a microservice:

type MyMicroservice struct {
	AgentBase
	printVersion *bool
}

func (m *MyMicroservice) AddAgentSpecificCLIFlags(flagSet *flag.FlagSet) {
	fmt.Println("MyMicroservice: AddAgentSpecificCLIFlags")
	m.printVersion = flagSet.Bool("v", false, "Print Version of the agent.")
}

// ProcessAgentSpecificCLIFlags is not needed for MyMicroservice
// Also there is never a need to reimplement GetAgentBase

func Run(args []string) {
	m := &MyMicroservice{}
	_ = Init(m, "my-agent", WithArguments(args))
	if *m.printVersion {
		fmt.Println("My version is...")
		os.Exit(0)
	}
	// Run microservice ...
}

func TestAgentBase(test *testing.T) {
	Run([]string{"-v"})
}
```

I also tried to avoid the word "Context", which is not very Go idiomatic.

Reply:
@milan-zededa thank you for the great suggestion. This way we can minimize code change and avoid duplicates. Done.
I modified *Context to *AgentState in code additions, but decided to not touch existing naming in the PR as it will enormous enlarge the diff. Will rename in another PR.

Reply:
@eriknordmark Best I could come up with was to reuse package name for that - AgentBase. But AgentConfig would be probably fine as well. The thing is thought that this structure contains also logger and pubsub, neither of which is config or status.

Reply:
@giggsoff Glad that you found my proposal useful. And yes let's not rename existing "Contexts" here, this was just a suggestion going forward.